Ensure `filename` in xopp format is always saved as a relative path to the directory it resides in

hyperupcall opened the pull request:
There are several issues stemming from the fact that annotated PDF backgrounds break when:

1. Renaming / moving containing directories because `filename` is stored as an absolute path
2. Saving a `.xopp` file different than the `$PWD` when opening one from the command line (passing it as $1)
  - This bug doesn't surface in the common case because xournal is usually launched through a launcher, where the `$PWD` is equivalent to `$HOME`, which is nearly always the common parent of xopp and pdf files.
  
Specifically, this
- Closes #3880 (relative path is now relative with respect to the newly created `.xopp` file, rather than `$PWD`)
- Closes #3273 (this particular issue is a specialization of the previous issue)
- Closes #2736 (relative paths are guaranteed to be generated, since we call `std::filesystem::relative` with two absolute paths, which uses [`lexically_relative`](https://en.cppreference.com/w/cpp/filesystem/path/lexically_normal) under the hood after canonicalization
- Closes #2638 (same as previous, the issue seems stale too)
- Closes #2774 (see [comment](https://github.com/xournalpp/xournalpp/pull/4262#issuecomment-1408067035))

One of the issues floated around [this changeset](https://github.com/emorice/xournalpp/commit/6bdc96d427a6d68acafb54c815d6e8abe9f83c83), which I unfortunately found only after I did my on work - which appears to do nearly the same thing except:
- I applied my changes to not just background PDFs, but for images too. (wherever I saw a `background->setAttrib("domain", "absolute")` near a `background->setAttrib("filename", ...)` in the save handler, I performed this change)

Validation method:
- manual testing

bhennion commented on src/core/control/jobs/AutosaveJob.cpp:
```diff
@@ -30,10 +30,7 @@ void AutosaveJob::run() {
 
-    doc->lock();
-    handler.prepareSave(doc);
     auto filepath = doc->getFilepath();
```
I think `doc->getFilepath();` still needs the mutex lock (imagine the user renames the file -- or saves as -- while the autosave fetchs the filepath.

bhennion commented on src/core/control/jobs/SaveJob.cpp:
```diff
@@ -98,13 +98,14 @@ auto SaveJob::save() -> bool {
 
-    doc->lock();
-    h.prepareSave(doc);
     fs::path filepath = doc->getFilepath();
```
Same remark here

bhennion reviewed (changes requested):
Only one mutex related comment. The rest LGTM.

Ideally, this should be tested on various platforms, on shared partitions (e.g. in dual boot), but also on external media (usb key, external drive), and on cloud sync'd folders (e.g. Dropbox) opened on different computers.

bhennion commented:
@rolandlo If I remember correctly, you had worked on the attached pdf mode back in the day. Would you mind giving this a glance in case there was a hidden reason for absolute paths I'm not aware of?

rolandlo commented:
> @rolandlo If I remember correctly, you had worked on the attached pdf mode back in the day. Would you mind giving this a glance in case there was a hidden reason for absolute paths I'm not aware of?

I will take a look at it. There absolutely might be a hidden reason for absolute paths, but I'm not aware of it right now and I will have to test in various situations.

Technius commented:
I think there's one use case for absolute paths, which is moving the `.xopp` file but leaving the `.pdf` file in the same place and only using these annotations on the same computer. Otherwise, I think relative paths are preferred in all other scenarios.

This whole relative/absolute path problem can be entirely avoided with the packaged file format, but I've had almost no time to work on it in the past few months.

hyperupcall commented:
Cool beans, got that locking fixed!

@Technius Yeah, that's the only use case I can think of for absolute paths too. I figured it's much less intuitive and is extremely fragile. For example, if a set of files is on a USB, it could break when switching between different Linux distros.

rolandlo commented:
Does that really work on Windows when the PDF (or image) and the `xopp`-file are on different drives?
Seems @Febbe has some experience, see [this question](https://stackoverflow.com/questions/63710574/stdfilesystemrelative-bug-in-mingw-or-msvc)
In general since the implementation of `std::filesystem` is operating system dependent, a lot of care and tests should be made before stepping ahead with this PR.

hyperupcall commented:
OMG!! MSVC returns an empty string? Yeah let's hold this - it looks like we have to give special treatment to msvc and cygwin or something...

Just to clarify, we are supporting cygwin/mingw or mysys2 as a platform? I usually forget Windows, so I'm not sure if to treat those as separate platforms or lump some of them together, or which ones are most common?

hyperupcall commented:
Yeah - I was thinking we can have an exception for Windows - where if the drive letter is different, use the absolute path. I suppose this would mean we would have to call a helper function, and do a little dirty work in there just for Windows - does that sound okay?

With some of my other PRs merged, I think this is now something I can consider implementing

Technius commented:
Given that the file name / path conversions have been a source of numerous bugs over the years, it would be nice if we could also refactor the path resolution logic into helper functions, and then write unit tests for them.

One potential solution for dealing with the separate drive problem is to only save as a relative path when their [root paths](https://en.cppreference.com/w/cpp/filesystem/path/root_path) are the same.

Febbe commented:
> So... I'm not sure what to do with this. I guess we should treat differently the case were the .xopp file and the .pdf file are not on the same partition. I don't know if this can be easily achieved using std::filesystem.

That should be achievable via std::filesystem
Build relative path, when that fails with empty path or exception use absolute path.

@Technius it might be better to directly try to calculate a relative path. When it does not exist or is unresolvable then use the canonical path.

hyperupcall commented:
The latest commit addresses feedback; namely:

- If the drive letters are different on Windows, then simply use the absolute path (see test)
- If the final calculated "relative path" is an empty string, then simply use the absolute path
  - This check was added so that if there are any other path bugs with the filesystem library in the future, it will use the safe path of using the absolute path. I didn't think this was a _full_ replacement for the first `.root_path()` check though. For example, if the filesystem library somehow fixes the bug and calculates the path correctly, I wanted to ensure that the behavior stays consistent by still using absolute paths

hyperupcall commented:
I just fixed up the PR again, renaming the method to `normalizeAssetPath`, and changing the following:

On Windows, the path is now normalized to use UNIX-style file separators. I checked that these paths are still serializable to `fs::path` on Windows. This was done to make the data in the file format more cross-platform and it additionally partially address #3412. By making the path separators the same, simply saving the file on a different platform won't needlessly clutter the diff.

Additional tests were also added to ensuer the saved path is non-empty (which is a possibility on windows when `.root_path()` is different, or when the first argument to `std::relative` is not an absolute path.

danemtsov commented on src/util/PathUtil.cpp:
```diff
@@ -270,6 +270,30 @@ auto Util::ensureFolderExists(const fs::path& p) -> fs::path {
 }
 
+auto Util::normalizeAssetPath(const fs::path& asset_path, const fs::path& base) -> std::string {
+    auto final_path = std::string{};
+
+#ifdef _WIN32
+    // On Windows, fs::relative may return an empty or malformed path if the
+    // the root_path()s are different
+    if (asset_path.root_path() == base.root_path()) {
+        final_path = fs::relative(asset_path, base).string();
+        std::replace(final_path.begin(), final_path.end(), '\\', '/');
+    }
+#else
+    // With other platforms, the root is always '/', so a relative path can
+    // always be correctly constructed
+    final_path = fs::relative(asset_path, base).string();
+#endif
+
+    // In some cases (if `asset_path` is a relative path), `final_path` may be
+    // empty, so ensure that a valid path is always returned from this function
+    if (final_path.empty())
+        final_path = fs::weakly_canonical(asset_path).string();
+
+    return final_path;
```
[path::generic_string() and friends](https://en.cppreference.com/w/cpp/filesystem/path/generic_string) does precisely what we want here. IMO we should always use this (and its counterpart, the constructor with `format::generic_format`) when dealing with file I/O.

```suggestion
    // fs::relative() may return an empty or malformed path if the root_path()s are different
    return (asset_path.root_path() == base.root_path() ?
            fs::relative(asset_path, base) : fs::weakly_canonical(asset_path)).generic_u8string();
```

danemtsov commented on test/unit_tests/util/PathTest.cpp:
```diff
@@ -117,3 +117,35 @@ TEST(UtilPath, testClearExtensions) {
+    auto p = fs::path();
+
+#if _WIN32
+    p = Util::normalizeAssetPath("C:\\dir\\file.txt", "D:");
+    EXPECT_EQ(string("C:\\dir\\file.txt"), p.string());
+
+    p = Util::normalizeAssetPath("C:\\dir\\file.txt", "D:\\base");
+    EXPECT_EQ(string("C:\\dir\\file.txt"), p.string());
+
+    // do not return empty if asset_path is relative
+    p = Util::normalizeAssetPath("../dir/file.txt", "D:");
+    EXPECT_TRUE(!p.empty());
+
+    p = Util::normalizeAssetPath("../dir/file.txt", "D:\\base");
+    EXPECT_TRUE(!p.empty());
```
We'll want the result to always have forward slashes.

danemtsov commented on src/core/control/xojfile/SaveHandler.cpp:
```diff
@@ -251,7 +251,8 @@ void SaveHandler::visitPage(XmlNode* root, PageRef p, Document* doc, int id) {
             } else {
                 background->setAttrib("domain", "absolute");
-                background->setAttrib("filename", doc->getPdfFilepath().string());
+                auto normalizedPath = Util::normalizeAssetPath(doc->getPdfFilepath().string(), target.parent_path());
```
```suggestion
                auto normalizedPath = Util::normalizeAssetPath(doc->getPdfFilepath(), target.parent_path());
```

danemtsov commented on src/core/control/xojfile/SaveHandler.cpp:
```diff
@@ -276,7 +277,10 @@ void SaveHandler::visitPage(XmlNode* root, PageRef p, Document* doc, int id) {
         } else {
             background->setAttrib("domain", "absolute");
-            background->setAttrib("filename", p->getBackgroundImage().getFilepath().string());
+            auto normalizedPath = Util::normalizeAssetPath(
+                    std::filesystem::current_path() / p->getBackgroundImage().getFilepath().string(),
+                    target.parent_path());
```
```suggestion
            auto normalizedPath = Util::normalizeAssetPath(
                    p->getBackgroundImage().getFilepath(),
                    target.parent_path());
```

danemtsov reviewed (commented):
I think it's better to check the root path even for non-Windows platforms (see [cppreference](https://en.cppreference.com/w/cpp/filesystem/path) on path syntax) - better safe than sorry.
Plus I don't think it makes sense to store a relative path if the locations are that 'far apart' (i.e. not even same root path).

https://github.com/xournalpp/xournalpp/pull/4262/files#diff-132c5a6df123d556477b8e79c5e5154c83a5075ef45bb9feadae8b46cac01428R281

Probably better to ensure all paths are absolute?

BTW we must also ensure all the strings are converted as UTF-8 to avoid locale and issues.

Febbe commented:
@hyperupcall this PR is nearly done and merge-able, actually only the tests must be corrected. So my question is, if you still want to continue working on it and if you have time to do it?

If you are pretty busy, we would just finish it.

hyperupcall commented:
If I don't push any commits by April 13th, feel free to finish it up. I wish I was able to close this earlier

Febbe commented:
> Probably better to ensure all paths are absolute?

A heuristic, and the ability, to manage assets afterward would be good. Neither completely absolute nor relative paths are a good final solution. 
But we can implement the heuristic and the asset manager later.

Febbe commented on test/unit_tests/util/PathTest.cpp:
```diff
@@ -10,6 +10,8 @@
  */
 
 
+#include <filesystem>
+
```
That should be reverted, or we should choose, to abandon compilers / STL implementations which do not provide `std::filesystem`

Febbe commented:
@xournalpp/core before we merge this, we should discuss, if we really wan't to override all old files.
 - I think in all loaded files, asset paths should be saved the same as they are loaded before.
 - when a user relocates a xournal file via "save as" he should be asked, if the relative paths should be updated, or if he want to keep them.
 - new assets should be added with an relative path

Technius commented:
Let's save this for 1.3.0. I don't think we should be making such a large change right before the next release.

KlausBlum commented:
+1 for relative paths to background PDFs!  
  
This would solve lots or problems. It would be cool to have an option for relative or absolute paths as default setting.

Are there any reasons this PR is still waiting?   
Putting it into a nightly build would enable us to do all kinds of tests. :-)
